NLP/main/extract_display_quotes.py

```diff
@@ -0,0 +1,531 @@
+                                        write_tree=create_tree, 
+                                        tree_dir=tree_dir)
+                
+                # extract the included named entities
+                speaks, qts = [quote['speaker'] for quote in quotes], [quote['quote'] for quote in quotes]        
+                speak_ents = [[(str(ent), ent.label_) for ent in doc.ents if (str(ent) in speak) & (ent.label_ in inc_ent)] for speak in speaks]        
+                quote_ents = [[(str(ent), ent.label_) for ent in doc.ents if (str(ent) in qt) & (ent.label_ in inc_ent)] for qt in qts]
+        
+                # add text_id, quote_id and named entities to each quote
+                for n, quote in enumerate(quotes):
+                    quote['text_id'] = doc_id
+                    quote['quote_id'] = str(n)
+                    quote['speaker_entities'] = list(set(speak_ents[n]))
+                    quote['quote_entities'] = list(set(quote_ents[n]))
+                    
+                # store them in all_quotes
+                all_quotes.extend(quotes)
+                    
+            except:
+                # this will provide some information in the case of an error
+                self.app_logger.exception("message")
+                traceback.print_exception()
+                
+        # convert the outcome into a pandas dataframe
+        quotes_df = pd.DataFrame.from_dict(all_quotes)
+        
+        # convert the string format quote spans in the index columns to a tuple of integers
+        for column in quotes_df.columns:
+            if column.endswith('_index'):
+                quotes_df[column].replace('','(0,0)', inplace=True)
+                quotes_df[column] = quotes_df[column].apply(eval)
+        
+        # re-arrange the columns
+        new_index = ['text_id', 'quote_id', 'quote', 'quote_index', 'quote_entities', 
+                     'speaker', 'speaker_index', 'speaker_entities',
+                     'verb', 'verb_index', 'quote_token_count', 'quote_type', 'is_floating_quote']
+        quotes_df = quotes_df.reindex(columns=new_index)
+                
+        return quotes_df
+    
+    
+    def show_quotes(self, text_df, quotes_df, text_id, show_what, inc_ent, save_to_html=False):
+        '''
+        Display speakers, quotes and named entities inside the text using displaCy
+
+        Args:
+            text_df: the pandas dataframe containing the list of texts
+            quotes_df: the pandas dataframe containing the quotes
+            text_id: the text_id of the text you wish to display
+            show_what: options to display speakers, quotes or named entities
+            inc_ent: a list containing the named entities to be extracted from the text, 
+                     e.g., ['ORG','PERSON','GPE','NORP','FAC','LOC']
+            save_to_html: option to save the display into an html file
+        '''
+        # formatting options
```

Comment:
Formatting options can also be considered to be loaded/formed from a yml file in a later stage. Good for now.

Reply:
will modify later - would be good if u could give me an example